src/mca/gds/base/gds_base_fns.c

```diff
@@ -210,24 +259,120 @@ pmix_status_t pmix_gds_base_store_modex(struct pmix_namespace_t *nspace,
                     &bkt, &bo2, &cnt, PMIX_BYTE_OBJECT);
         }
         PMIX_DESTRUCT(&bkt);
+
         if (PMIX_ERR_UNPACK_READ_PAST_END_OF_BUFFER == rc) {
             rc = PMIX_SUCCESS;
         } else if (PMIX_SUCCESS != rc) {
-            goto error;
+            PMIX_ERROR_LOG(rc);
+            goto exit;
         }
         /* unpack and process the next blob */
         cnt = 1;
         PMIX_BFROPS_UNPACK(rc, pmix_globals.mypeer,
                 buff, &bo, &cnt, PMIX_BYTE_OBJECT);
     }
+
     if (PMIX_ERR_UNPACK_READ_PAST_END_OF_BUFFER == rc) {
         rc = PMIX_SUCCESS;
+    } else if (PMIX_SUCCESS != rc) {
+        PMIX_ERROR_LOG(rc);
+    }
+exit:
+    if (pmix_argv_count(kmap) > 0) {
+        pmix_argv_free(kmap);
+    }
+    return rc;
+}
+
+/*
+ * Pack the key-value as a tuple of key-name index and key-value.
+ * The key-name to store replaced by unique key-index that stored
+ * to the key-map. So the remote server can determine the key-name
+ * by the index from map that packed in modex as well.
+ *
+ * kmap - key values array by (char*), uses to store unique key
+ *        names string and determine their indexes
+ *
+ * buf - output buffer to pack key-values
+ *
+ * kv - pmix key-value pair
+ */
+pmix_status_t pmix_gds_base_modex_pack_kval(char ***kmap, pmix_buffer_t *buf,
+                                            pmix_kval_t *kv)
+{
+    bool found = false;
+    uint32_t key_idx;
+    char *tmp;
+    pmix_status_t rc = PMIX_SUCCESS;
+    int size = pmix_argv_count(*kmap);
+
+    /* determine the uniq index of key */
```

Comment:
We have a pmix_argv_append_unique_nosize function that will do this for you. Only problem is that it returns "success" instead of the index where it put or found the string. We wouldn't want to modify that function as it gets used in various places, but maybe we should add another one (e.g., pmix_argv_append_unique_idx) that does the same logic but returns the index of the stored location?

Reply:
`pmix_argv_append_unique_idx` function added in the separate commit and updated to use here.